modules/drivers/redshift/test/metabase/driver/redshift_test.clj

```diff
@@ -391,3 +394,117 @@
+                       "order by start_time desc\n"
+                       "limit 5\n")]
+    (->> (run-native-query query-str)
+         mt/rows
+         (map (fn [[session-id start-time result-cache-hit query-text]]
+                {:session-id session-id
+                 :start-time start-time
+                 :result-cache-hit result-cache-hit
+                 :query-text query-text})))))
+
+;; According to https://docs.aws.amazon.com/redshift/latest/dg/c_challenges_achieving_high_performance_queries.html#result-caching
+;; caching is on by default. Hence check for that is skipped.
+(deftest server-side-timestamp-caching-test
+  (mt/test-driver :redshift
+   (testing "Relative date time queries should be cached"
+     (sql-jdbc.execute/do-with-connection-with-options
+      :redshift
+      (mt/id)
+      nil
+      (fn [^java.sql.Connection _connection]
+        (let [sid (session-id!)
+              ;; :time-interval translates to :relative-datetime, which is now using server side generated timestamps
+              ;; for units day and greater.
+              query-to-cache (mt/mbql-query
+                              test_data_users
+                              {:fields [$id $last_login]
+                               :filter [:time-interval $last_login -500 :day {:include-current false}]})
+              compiled-sql (-> query-to-cache qp/compile :query)]
+          (dotimes [_ 2]
+            (qp/process-query query-to-cache))
+          (let [last-queries (last-queries-of-a-session sid)
+                ;; First row of last queries should contains "last-queries query" info.
+                ;; Second should contain info about query that should be returned from cache.
+                cached-row (second last-queries)
+                cached-row-adjusted-redshift-sql (str/replace (:query-text cached-row) #"\$\d+" "?")
+                ;; Third row contains the same query. At the point of its execution, it may have been cached, but
+                ;; afterwards it should.
+                pre-cached-row (nth last-queries 2)
+                pre-cached-row-adjusted-redshift-sql (str/replace (:query-text pre-cached-row) #"\$\d+" "?")]
+            (testing "Last queries results contain query before caching on expected postion"
+              (is (.contains pre-cached-row-adjusted-redshift-sql compiled-sql)))
+            (testing "Last queries results contain cached query on expected position"
+              (is (.contains cached-row-adjusted-redshift-sql compiled-sql)))
+            (testing "Query was returned from cache"
+              (is (true? (:result-cache-hit cached-row)))))))))))
+
+(defn getdate-vs-ss-ts-test-thunk-generator
+  ([]
+   (getdate-vs-ss-ts-test-thunk-generator :week -1))
+  ([unit value]
+   (fn []
+     (let [honey {:select [[(with-redefs-fn {#'redshift/use-server-side-timestamp? (constantly false)}
+                              #(sql.qp/->honeysql :redshift [:relative-datetime value unit]))]
```

Comment:
This looks suspicious. When does the lambda run? It seems to escape the scope of the `with-redefs-fn` and run with the restored settings. Either the lambda has to be removed or `with-redefs-fn` has to be moved inside it.

Are you playing with the lambda to make it less likely that the two expressions are executed on different days? If so, it might be easier to skip the test in the "danger zone".

Reply:
I've modified the code to use just `with-redefs`, not the `with-redefs-fn` variant.